plugin/cancel/cancel.py

```diff
@@ -1,20 +1,70 @@
 # to disable command cancellation, comment out this entire file.
 # you may also wish to adjust the commands in misc/cancel.talon.
-from talon import actions, speech_system
+
+import time
+
+from talon import Context, Module, actions, speech_system
+from talon.grammar import Phrase
 
 # To change the phrase used to cancel commands, you must also adjust misc/cancel.talon
 cancel_phrase = "cancel cancel".split()
 
+mod = Module()
+ctx = Context()
+
+ts_threshold: float = 0
+
+
+@ctx.action_class("speech")
+class SpeechActions:
+    # When Talon wakes we set the timestamp threshold. On the next command we
+    # will compare the phrase timestamp to the threshold and cancel any phrase
+    # started before wakeup. This is to prevent speech said before wake-up to
+    # be interpreted as a command if the user wakes Talon using a noise or
+    # keypress.
+    def enable():
+        actions.user.cancel_current_phrase()
+        actions.next()
+
+
+@mod.action_class
+class Actions:
+    def cancel_current_phrase():
+        """Cancel/abort current spoken phrase"""
+        global ts_threshold
+        ts_threshold = time.perf_counter()
 
-def pre_phrase(d):
-    n = len(cancel_phrase)
-    if "text" in d and "parsed" in d:
-        before, after = d["text"][:-n], d["text"][-n:]
-        if after != cancel_phrase:
+
+def pre_phrase(phrase: Phrase):
+    global ts_threshold
+
+    # Nothing to cancel
+    if "parsed" not in phrase:
+        return
+
+    # Check if the phrase is before the threshold
+    if ts_threshold != 0:
+        words = phrase["phrase"]
+        start = getattr(words[0], "start", phrase["_ts"])
```

Comment:
Note `words[0]` may not exist — just got this in my version:

```
    2: user/talon_community/core/modes/abort.py:32  | start = getattr(words[0], "start", phrase["_ts"])
    1:                                                                ^^^^^^^^
IndexError: list index out of range

```

Reply:
Good catch I have that in my own code, but in a earlier part of the call stack so I missed it.

Reply:
fixed

Reply:
Thanks. (Python's := operator can be useful in such cases but it's just a stylistic choice…)

Reply:
I did some refactoring and I think this turned out quite okay